homeassistant/components/nasweb/coordinator.py

```diff
@@ -0,0 +1,199 @@
+"""Message routing coordinators for handling NASweb push notifications."""
+from __future__ import annotations
+
+import asyncio
+from collections.abc import Callable
+from datetime import datetime, timedelta
+import logging
+import time
+from typing import Any
+
+from aiohttp.web import Request, Response
+from webio_api import Output as NASwebOutput, WebioAPI
+from webio_api.const import KEY_DEVICE_SERIAL, KEY_OUTPUTS, KEY_TYPE, TYPE_STATUS_UPDATE
+
+from homeassistant.const import Platform
+from homeassistant.core import CALLBACK_TYPE, HassJob, HomeAssistant, callback
+from homeassistant.helpers import event
+from homeassistant.helpers.entity_platform import async_get_platforms
+from homeassistant.helpers.update_coordinator import BaseDataUpdateCoordinatorProtocol
+
+from .const import DOMAIN, STATUS_UPDATE_MAX_TIME_INTERVAL
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class NotificationCoordinator:
+    """Coordinator redirecting push notifications for this integration to appropriate NASwebCoordinator."""
+
+    def __init__(self) -> None:
+        """Initialize coordinator."""
+        self._coordinators: dict[str, NASwebCoordinator] = {}
+
+    def add_coordinator(self, serial: str, coordinator: NASwebCoordinator) -> None:
+        """Add NASwebCoordinator to possible notification targets."""
+        self._coordinators[serial] = coordinator
+        _LOGGER.debug("Added NASwebCoordinator for NASweb[%s]", serial)
+
+    def remove_coordinator(self, serial: str) -> None:
+        """Remove NASwebCoordinator from possible notification targets."""
+        self._coordinators.pop(serial)
+        _LOGGER.debug("Removed NASwebCoordinator for NASweb[%s]", serial)
+
+    def has_coordinators(self) -> bool:
+        """Check if there is any registered coordinator for push notifications."""
+        return len(self._coordinators) > 0
+
+    async def check_connection(self, serial: str) -> bool:
+        """Wait for first status update to confirm connection with NASweb."""
+        nasweb_coordinator = self._coordinators.get(serial)
+        if nasweb_coordinator is None:
+            _LOGGER.error("Cannot check connection. No device match serial number")
+            return False
+        counter = 0
+        _LOGGER.debug("Checking connection with: %s", serial)
+        while not nasweb_coordinator.is_connection_confirmed() and counter < 10:
+            await asyncio.sleep(1)
+            counter += 1
+            _LOGGER.debug("Checking connection with: %s (%s)", serial, counter)
+        return nasweb_coordinator.is_connection_confirmed()
```

Comment:
This is a bit more Pythonic
```suggestion
        for counter in range(10):
            _LOGGER.debug("Checking connection with: %s (%s)", serial, counter)
            if nasweb_coordinator.is_connection_confirmed():
                return True
            await asyncio.sleep(1)
        return False
```